Source/DafnyLanguageServer/Plugins/QuickFixerHelpers.cs

```diff
@@ -0,0 +1,153 @@
+    var indentation = 0;
+    var indentationBrace = endToken.col - 1;
+    var firstNewline = true;
+    var useTabs = false;
+    // Look for the first newline
+    while (0 <= pos && pos < text.Length && (text[pos] != '\n' || firstNewline)) {
+      if (text[pos] == '\t') {
+        useTabs = true;
+      }
+
+      if (text[pos] == '\n') {
+        if (!firstNewline) {
+          break;
+        }
+
+        firstNewline = false;
+      }
+
+      if (!firstNewline) {
+        if (text[pos] == ' ' || text[pos] == '\t') {
+          indentation++;
+        } else {
+          indentation = 0;
+        }
+      }
+
+      pos--;
+    }
+
+    if (startToken.line == endToken.line - 1) {
+      // Override case {\n} with some spacing, we return a default value of 2 spaces or 1 tab, if we find some
+      indentation = indentationBrace + (useTabs ? 1 : 2);
+    } else if (startToken.line == endToken.line) {
+      // Override case { ... } on one line.
+      indentationBrace = startToken.col - 1;
+      indentation = indentationBrace + 1;
+    }
+
+    indentation = Math.Max(indentationBrace, indentation);
+
+    return (
+      new string(useTabs ? '\t' : ' ', indentation - indentationBrace),
+      new string(useTabs ? '\t' : ' ', indentationBrace));
+  }
+
+  /// <summary>
+  /// Given an opening brace (typically where an error is detected),
+  /// finds and returns the closing brace token.
+  /// </summary>
+  /// <param name="program">The program</param>
+  /// <param name="documentUri">The document URI</param>
+  /// <param name="openingBrace">A token whose pos is the absolute position of the opening brace</param>
+  /// <returns>The token of a matching closing brace, typically the `ÈndTok` of a BlockStmt</returns>
+  public static IToken? GetMatchingEndToken(Dafny.Program program, string documentUri, IToken openingBrace) {
+    // Look in methods for BlockStmt with the IToken as opening brace
+    // Return the EndTok of them.
+    foreach (var module in program.Modules()) {
+      foreach (var topLevelDecl in module.TopLevelDecls) {
+        if (topLevelDecl is ClassDecl classDecl) {
+          foreach (var member in classDecl.Members) {
+            if (member is Method method && method.tok.filename == documentUri && method.Body != null &&
+                GetMatchingEndToken(openingBrace, method.Body) is { } token) {
+              return token;
+            }
+
+            if (member is Function { ByMethodBody: { } } function &&
+                GetMatchingEndToken(openingBrace, function.ByMethodBody) is { } token2) {
+              return token2;
+            }
+          }
+        }
+      }
+    }
+
+    return null;
+  }
+
+  /// <summary>
+  /// Given an opening brace and a statement, if the statement's token is openingBrace
+  /// returns the closing brace token, else null.
+  /// Visit sub-statements recursively
+  /// </summary>
+  public static IToken? GetMatchingEndToken(IToken openingBrace, Statement stmt) {
```

Comment:
Is it possible to implement this without the `stmt` parameter?

Reply:
I don't think so. This method is recursive to traverse all statements. I could also use a visitor there, but I don't think it's necessary.